bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/Clip2Bridge.java

```diff
@@ -1039,14 +1040,16 @@ public void testConnectionState() throws HttpUnauthorizedException, ApiException
     }
 
     /**
-     * Hue Bridges get confused if they receive too many HTTP requests in a short period of time (e.g. on start up), or
-     * if too many HTTP sessions are opened at the same time. So this method throttles the requests to a maximum of one
-     * per REQUEST_INTERVAL_MILLISECS, and ensures that no more than MAX_CONCURRENT_SESSIONS sessions are started.
+     * The Hue Bridge can get confused if they receive too many HTTP requests in a short period of time (e.g. on start
+     * up), or if too many HTTP sessions are opened at the same time, which cause it to respond with an HTML error page.
+     * So this method throttles the requests to a maximum of one per REQUEST_INTERVAL_MILLISECS, and ensures that no
+     * more than MAX_CONCURRENT_SESSIONS stream permits are issued.
      *
+     * @param permitCount indicates how many stream permits to be acquired.
      * @throws InterruptedException
      */
-    private synchronized void throttle() throws InterruptedException {
-        streamMutex.acquire();
+    private synchronized void throttle(int permitCount) throws InterruptedException {
+        streamMutex.acquire(permitCount);
```

Comment:
I wonder whether the mutex/permit count combination should be a separate object, initialized in throttle() and released in throttleDone() so that the additional parameter in throttleDone() isn't needed and we can be sure to not have a a mismatch there.

Reply:
What I was heading at is the argument to throttle() and throttleDone(), which must match to work correctly. This kind of redundancy is at least potential for bugs.
My idea was to bundle the mutex and the count in some sort of holder object, so that upon releasing the mutex the count is readily available. Bonus points for making the holder object implement `AutoCloseable` and using try-with-resources for acquiring the lock ;-)

Reply:
Something like this (compile tested only): https://github.com/maniac103/openhab2-addons/commit/e88bac443b7f2f3e22d9626d799a292177816f3d

Reply:
> Bonus points for making the holder object implement AutoCloseable and using try-with-resources for acquiring the lock

:)

> Something like this

Ok. Looks good. Except that the `if (delay > 0) { Thread.sleep(delay); }` code needs to be inside the `synchronized` block.

But let me test it as well.

Reply:
Ok. Done. With slight adaption to your code.